### PR TITLE
12.0 custom selection option

### PR DESCRIPTION
**Pull request by ddufresne**
https://numigi.com/web#id=19745&action=298&model=project.task&view_type=form&menu_id=200

### CODE REVIEW

**Comment by ddufresne** [comment]
@foutoucour there is not a lot of code added, but the code was reorganised (one file per model).

File: web_custom_label/models/web_custom_label.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    @api.onchange('position')
+    def _onchange_position(self):
+        if self.position != 'selection':
+            self.key = False
+
+    @api.model
+    def create(self, vals):
+        new_record = super().create(vals)
+        modules.registry.Registry(self.env.cr.dbname).clear_caches()
```
Comment (foutoucour) [inline web_custom_label/models/web_custom_label.py:46]: pourquoi les `clear_caches` ?

File: web_custom_label/models/web_custom_label.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# © 2018 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
```
Comment (foutoucour) [inline web_custom_label/models/web_custom_label.py]:
```suggestion
# © 2018 - Today Numigi (tm) and all its contributors (https://bit.ly/numigiens)
```

File: web_custom_label/web_custom_label.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# © 2018 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
```
Comment (foutoucour) [inline web_custom_label/web_custom_label.py]: Fichier a enlevé je pense

File: web_custom_label/models/web_custom_label.py (hunk 1 above)
Comment (ddufresne) [reply web_custom_label/models/web_custom_label.py:46]: @foutoucour pour que la cache des vues et fields soient invalidées. De cette manière, je suis certain que si l'utilisateur raffraîchis son écran, les vues contiendront les nouveaux libellés.

File: web_custom_label/web_custom_label.py (hunk 3 above)
Comment (ddufresne) [reply web_custom_label/web_custom_label.py]: ah oui, une erreur de copier coller. Merci

File: web_custom_label/models/web_custom_label.py (hunk 1 above)
Comment (foutoucour) [reply web_custom_label/models/web_custom_label.py:46]: ok. Est ce que tu peux rajouter le commentaire dans le code ?

Comment (ddufresne) [reply web_custom_label/models/web_custom_label.py:46]: oui

Comment (ddufresne) [reply web_custom_label/models/web_custom_label.py:46]: @foutoucour done

File: web_custom_label/models/web_custom_label.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# © 2018 - Today Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from lxml import etree
+from typing import List, Mapping
```
Comment (ddufresne) [inline web_custom_label/models/web_custom_label.py]: Codacy Issue found: [Unused Mapping imported from typing](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

File: web_custom_label/models/ir_ui_view.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        arch_with_custom_labels = _add_custom_labels_to_view_arch(labels, arch)
+        set_custom_labels_on_fields(labels, fields)
+        return arch_with_custom_labels, fields
+
+
+def _add_custom_labels_to_view_arch(labels, arch):
+    if not labels:
+        return arch
+
+    tree = etree.fromstring(arch)
```
Comment (ddufresne) [inline web_custom_label/models/ir_ui_view.py]: Codacy Issue found: [Using lxml.etree.fromstring to parse untrusted XML data is known to be vulnerable to XML attacks. Replace lxml.etree.fromstring with its defusedxml equivalent function.](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

File: web_custom_label/models/ir_ui_view.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# © 2018 - today Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from lxml import etree
```
Comment (ddufresne) [inline web_custom_label/models/ir_ui_view.py]: Codacy Issue found: [Using etree to parse untrusted XML data is known to be vulnerable to XML attacks. Replace etree with the equivalent defusedxml package.](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

File: web_custom_label/models/web_custom_label.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# © 2018 - Today Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from lxml import etree
```
Comment (ddufresne) [inline web_custom_label/models/web_custom_label.py]: Codacy Issue found: [Unused etree imported from lxml](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

Comment (ddufresne) [inline web_custom_label/models/web_custom_label.py]: Codacy Issue found: [Using etree to parse untrusted XML data is known to be vulnerable to XML attacks. Replace etree with the equivalent defusedxml package.](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

**Review by foutoucour — APPROVED** [review]

File: web_custom_label/models/ir_ui_view.py (hunk 5 above)
Comment (ddufresne) [inline web_custom_label/models/ir_ui_view.py:31]: Codacy Issue found: [Using lxml.etree.fromstring to parse untrusted XML data is known to be vulnerable to XML attacks. Replace lxml.etree.fromstring with its defusedxml equivalent function.](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

File: web_custom_label/models/ir_ui_view.py (hunk 6 above)
Comment (ddufresne) [inline web_custom_label/models/ir_ui_view.py:4]: Codacy Issue found: [Using etree to parse untrusted XML data is known to be vulnerable to XML attacks. Replace etree with the equivalent defusedxml package.](https://app.codacy.com/manual/numigi/odoo-web-addons/pullRequest?prid=4960433)

**Pull request merged**